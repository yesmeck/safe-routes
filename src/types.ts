
import { type Preset } from '@react-router/dev/config';

export type ResolvedReactRouterConfig = Parameters<Required<Preset>['reactRouterConfigResolved']>[number]['reactRouterConfig'];
export type RequiredReactRouterConfig = Pick<ResolvedReactRouterConfig, 'appDirectory' | 'routes'>;
export type RouteManifestEntry = RequiredReactRouterConfig['routes'][string];
export type ReactRouterPluginContext = {
  reactRouterConfig: ResolvedReactRouterConfig;
}
