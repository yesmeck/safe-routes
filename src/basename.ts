export const placeholder = "__safe_routes_" + "basename__";

const basename = "__safe_routes_basename__";

export function resolveBasename(): string {
  if (basename && basename !== placeholder) {
    return basename;
  }
  return "";
}
