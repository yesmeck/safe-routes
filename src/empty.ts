export const empty = 0;
