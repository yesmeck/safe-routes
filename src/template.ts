interface Context {
  strict?: boolean;
  relativeAppDirPath: string;
  routes: Array<{
    id: string,
    route: string;
    params: string[];
    fileName: string;
  }>,
}

function indent(n: number) {
  return '  '.repeat(n);
}

function exportedQuery(ctx: Context) {
  if (ctx.strict) {
    return `type ExportedQuery<T> = IsSearchParams<T> extends true ? T : never;`;
  }
  return `type ExportedQuery<T> = IsSearchParams<T> extends true ? T : URLSearchParamsInit;`;
}

function routes(ctx: Context) {
  const routes = ctx.routes.map(({ route, params, fileName }) =>
    `"${route}": {
      id: '${route}',
      params: ${params.length > 0 ? `{${params.map(param => `${param}: string | number`).join('; ')}}` : 'never'},
      query: ExportedQuery<import('${ctx.relativeAppDirPath}/${fileName}').SearchParams>,
    }`
  );

  return `export interface Routes {
    ${routes.join(',\n' + indent(2))}
  }`
}

export function template(ctx: Context) {
  return `declare module "safe-routes" {
  type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
  // symbol won't be a key of SearchParams
  type IsSearchParams<T> = symbol extends keyof T ? false : true;

  ${exportedQuery(ctx)}

  ${routes(ctx)}

  type RoutesWithParams = Pick<
    Routes,
    {
      [K in keyof Routes]: Routes[K]["params"] extends Record<string, never> ? never : K
    }[keyof Routes]
  >;

  export type RouteId = Routes[keyof Routes]['id'];

  export function $path<
    Route extends keyof Routes,
    Rest extends {
      params: Routes[Route]["params"];
      query?: Routes[Route]["query"];
    }
  >(
    ...args: Rest["params"] extends Record<string, never>
      ? [route: Route, query?: Rest["query"]]
      : [route: Route, params: Rest["params"], query?: Rest["query"]]
  ): string;

  export function $params<
    Route extends keyof RoutesWithParams,
    Params extends RoutesWithParams[Route]["params"]
  >(
      route: Route,
      params: { readonly [key: string]: string | undefined }
  ): {[K in keyof Params]: string};

  export function $routeId(routeId: RouteId): RouteId;
}`;
};
