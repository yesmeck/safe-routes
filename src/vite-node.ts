import { createServer } from 'vite'
import { ViteNodeRunner } from 'vite-node/client'
import { ViteNodeServer } from 'vite-node/server'

export async function createContext() {
  const server = await createServer({
    server: {
      preTransformRequests: false,
      hmr: false,
    },
    optimizeDeps: {
      noDiscovery: true,
    },
  })

  await server.pluginContainer.buildStart({})

  const node = new ViteNodeServer(server)

  const runner = new ViteNodeRunner({
    root: server.config.root,
    base: server.config.base,
    fetchModule(id) {
      return node.fetchModule(id)
    },
    resolveId(id, importer) {
      return node.resolveId(id, importer)
    },
  })

  return { server, runner };
}
