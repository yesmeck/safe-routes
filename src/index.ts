import { resolveBasename } from './basename.js';

function includesParams(route: string) {
  return route.indexOf('/:') > -1 || route.indexOf('/*') > -1;
}

function prependBasename(path: string) {
  const basename = resolveBasename();
  if (!basename) {
    return path;
  }
  return basename + path;
}

export function $path(route: string, ...paramsOrQuery: Array<any>) {
  let path = route;
  let query:
    | string
    | string[][]
    | Record<string, string>
    | URLSearchParams
    | undefined = paramsOrQuery[0];

  if (includesParams(route)) {
    const params: any = paramsOrQuery[0] ?? {};
    query = paramsOrQuery[1];
    path = route.split('/').map(fragment => {
      if (fragment.indexOf('?') > -1) {
        fragment = fragment.slice(0, -1);
      }
      if (fragment.indexOf(':') > -1) {
        let [paramName, extension] = fragment.slice(1).split('.');
        if (paramName in params && params[paramName] !== undefined) {
          return params[paramName] + (extension ? '.' + extension : '');
        }
        return null
      }
      if (fragment == "*") {
        if ("*" in params) {
          return params["*"];
        }
        return null;
      }
      return fragment;
    }).filter(f => f !== null).join('/');
  }

  if (!query) {
    return prependBasename(path);
  }

  const searchParams = new URLSearchParams(query);


  return prependBasename(path + '?' + searchParams.toString());
}

export function $routeId(routeId: string) {
  return routeId;
}
