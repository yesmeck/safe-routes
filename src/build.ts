import * as fs from 'fs';
import { mkdirp } from 'mkdirp';
import * as path from 'path';
import slash from 'slash';
import { template } from './template.js';
import type { RequiredReactRouterConfig, RouteManifestEntry } from './types.js';

interface Options {
  strict?: boolean;
  outputDirPath: string;
}

type RoutesInfo = Record<string, {
  id: string;
  fileName: string;
  params: string[];
}>

export const DEFAULT_OUTPUT_DIR_PATH = './.react-router/types'

async function buildHelpers(config: RequiredReactRouterConfig): Promise<RoutesInfo> {
  const routesInfo: RoutesInfo = {};
  const handleRoutesRecursive = (
    parentId?: string,
    parentPath: RouteManifestEntry[] = [],
  ) => {
    let routes = Object.values(config.routes).filter(
      (route) => route.parentId === parentId,
    );
    routes.forEach((route) => {
      let currentPath = parentPath;
      if (route.id === 'root') {
        routesInfo['/'] = {
          id: route.id,
          fileName: route.file,
          params: [],
        };
      } else {
        currentPath = [...currentPath, route];
        const fullPath = dedupPrefixSlash(currentPath.reduce(
          (acc, curr) => [acc, trimSlash(curr.path)].filter(p => p != undefined).join('/'),
          '',
        ));
        const paramsNames = parse(currentPath);
        // Expand out the paths into all of the possible variants, taking into
        // account optional segments that aren't params/dynamic.
        for (const pathVariant of expandOptionalStaticSegments(fullPath)) {
          routesInfo[pathVariant] = {
            id: route.id,
            fileName: route.file,
            params: paramsNames
          };
        }
      }
      handleRoutesRecursive(route.id, currentPath);
    });
  };
  handleRoutesRecursive();
  return routesInfo;
}

function expandOptionalStaticSegments(path: string) {
  // Split the path at the point where `/` is the next char (positive lookahead regex)
  const segments = path.split(/(?=\/)/g);
  // This turns '/foo/bar' -> ['/foo', '/bar']
  let paths = [''];
  for (const e of segments) {
    if (!e.endsWith('?') || e.startsWith('/:')) {
      // If a given segment is not optional or dynamic, then we just append it to
      // each path variant we have so far: ['/foo'] => ['/foo/bar']
      paths = paths.map((p) => p + e);
    } else {
      // If a given segment is optional, we append a copy of the existing paths
      // with the optional segment appended:
      // ['/foo'] => ['/foo', '/foo/bar']
      paths.push(...paths.map((p) => p + e.slice(0, -1)));
    }
  }
  return paths;
}

export async function build(root: string, config: RequiredReactRouterConfig, options: Options) {
  const routesInfo = await buildHelpers(config);
  generate(root, config, routesInfo, options);
}

function generate(root: string, config: RequiredReactRouterConfig, routesInfo: RoutesInfo, options: Options) {
  const outputPath = path.join(
    root,
    options.outputDirPath,
  );
  const relativeAppDirPath = slash(path.relative(outputPath, config.appDirectory));
  const tsCode = template({
    strict: options.strict,
    relativeAppDirPath,
    routes: Object.entries(routesInfo).map(([route, { id, fileName, params }]) => ({
      id,
      route,
      params,
      fileName: slash(fileName.replace(/\.tsx?$/, '')),
    })).sort((a, b) => a.route.localeCompare(b.route)),
  });

  if (!fs.existsSync(outputPath)) {
    mkdirp.sync(outputPath);
  }
  fs.writeFileSync(path.join(outputPath, 'safe-routes.d.ts'), tsCode);
}

function parse(routes: RouteManifestEntry[]) {
  const paramNames: string[] = [];
  routes.forEach((route) => {
    return (
      route.path &&
      paramNames.push(
        ...route.path
          .split('/')
          .filter((seg) => seg.startsWith(':') || seg == '*')
          .map((param) => param.split('.')[0])
          .map((param) => param.replace(':', '').replace('*', '"*"')),
      )
    );
  });
  return paramNames;
}

function dedupPrefixSlash(path: string) {
  return path.replace(/^\/+/, '/');
}

function trimSlash(path?: string) {
  if (!path) return path;
  return path.replace(/\/+$/, '');
}
