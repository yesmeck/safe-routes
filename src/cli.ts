#!/usr/bin/env node
import path from 'node:path';
import { createContext } from './vite-node.js';

async function run() {
  const { runner, server } = await createContext();
  await runner.executeFile(path.resolve(import.meta.dirname, '../dist/empty.js'));
  await server.close();
  process.exit(0);
}

const command = process.argv[2];

if (!command || command !== 'typegen') {
  console.log('Usage: safe-routes typegen');
  process.exit(1);
}

run();
