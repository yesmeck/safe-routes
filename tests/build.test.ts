import * as fs from 'node:fs';
import * as os from 'node:os';
import * as path from 'node:path';
import { expect, test } from 'vitest';
import { build } from '../src/build';
import { testRoutes } from './fixture';

test('gen route types', async () => {
  const root = fs.mkdtempSync(path.join(os.tmpdir(), 'safe-routes-'));
  const config = {
    appDirectory: path.join(root, "app"),
    routes: testRoutes,
  }

  await build(root, config, { outputDirPath: './node_modules', strict: false });

  expect(
    fs.readFileSync(path.join(root, 'node_modules/safe-routes.d.ts'), 'utf8'),
  ).toMatchSnapshot();
});
