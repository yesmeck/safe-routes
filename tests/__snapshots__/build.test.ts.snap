// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`gen route types 1`] = `
"declare module "safe-routes" {
  type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
  // symbol won't be a key of SearchParams
  type IsSearchParams<T> = symbol extends keyof T ? false : true;

  type ExportedQuery<T> = IsSearchParams<T> extends true ? T : URLSearchParamsInit;

  export interface Routes {
    "/": {
      id: '/',
      params: never,
      query: ExportedQuery<import('../app/root').SearchParams>,
    },
    "/:lang?/about": {
      id: '/:lang?/about',
      params: {lang?: string | number},
      query: ExportedQuery<import('../app/routes/($lang).about').SearchParams>,
    },
    "/admin": {
      id: '/admin',
      params: never,
      query: ExportedQuery<import('../app/routes/admin._index').SearchParams>,
    },
    "/admin/episodes": {
      id: '/admin/episodes',
      params: never,
      query: ExportedQuery<import('../app/routes/admin.episodes._index').SearchParams>,
    },
    "/admin/episodes/:id": {
      id: '/admin/episodes/:id',
      params: {id: string | number},
      query: ExportedQuery<import('../app/routes/admin.episodes.$id._index').SearchParams>,
    },
    "/admin/episodes/:id/comments": {
      id: '/admin/episodes/:id/comments',
      params: {id: string | number},
      query: ExportedQuery<import('../app/routes/admin.episodes.$id.comments').SearchParams>,
    },
    "/admin/episodes/new": {
      id: '/admin/episodes/new',
      params: never,
      query: ExportedQuery<import('../app/routes/admin.episodes.new').SearchParams>,
    },
    "/api/:id.json": {
      id: '/api/:id.json',
      params: {id: string | number},
      query: ExportedQuery<import('../app/routes/api.$id[.]json').SearchParams>,
    },
    "/auth": {
      id: '/auth',
      params: never,
      query: ExportedQuery<import('../app/routes/auth._auth').SearchParams>,
    },
    "/auth/login": {
      id: '/auth/login',
      params: never,
      query: ExportedQuery<import('../app/routes/auth._auth.login').SearchParams>,
    },
    "/blog/rss.xml": {
      id: '/blog/rss.xml',
      params: never,
      query: ExportedQuery<import('../app/routes/blog.rss[.]xml').SearchParams>,
    },
    "/chats/:season/:episode": {
      id: '/chats/:season/:episode',
      params: {season: string | number; episode: string | number},
      query: ExportedQuery<import('../app/routes/chats_.$season.$episode').SearchParams>,
    },
    "/chats/:season/:episode/:slug": {
      id: '/chats/:season/:episode/:slug',
      params: {season: string | number; episode: string | number; slug: string | number},
      query: ExportedQuery<import('../app/routes/chats_.$season.$episode.$slug').SearchParams>,
    },
    "/credits": {
      id: '/credits',
      params: never,
      query: ExportedQuery<import('../app/routes/credits').SearchParams>,
    },
    "/home": {
      id: '/home',
      params: never,
      query: ExportedQuery<import('../app/routes/(static).home').SearchParams>,
    },
    "/jokes": {
      id: '/jokes',
      params: never,
      query: ExportedQuery<import('../app/routes/jokes._index').SearchParams>,
    },
    "/jokes/:jokeId": {
      id: '/jokes/:jokeId',
      params: {jokeId: string | number},
      query: ExportedQuery<import('../app/routes/jokes.$jokeId').SearchParams>,
    },
    "/people/:personId": {
      id: '/people/:personId',
      params: {personId: string | number},
      query: ExportedQuery<import('../app/routes/people.$personId').SearchParams>,
    },
    "/people/:personId/:planId/remove-plan": {
      id: '/people/:personId/:planId/remove-plan',
      params: {personId: string | number; planId: string | number},
      query: ExportedQuery<import('../app/routes/people.$personId.$planId.remove-plan').SearchParams>,
    },
    "/posts": {
      id: '/posts',
      params: never,
      query: ExportedQuery<import('../app/routes/posts._index').SearchParams>,
    },
    "/posts/:id": {
      id: '/posts/:id',
      params: {id: string | number},
      query: ExportedQuery<import('../app/routes/posts.$id').SearchParams>,
    },
    "/posts/*": {
      id: '/posts/*',
      params: {"*": string | number},
      query: ExportedQuery<import('../app/routes/posts.$').SearchParams>,
    },
    "/posts/delete": {
      id: '/posts/delete',
      params: never,
      query: ExportedQuery<import('../app/routes/posts.delete').SearchParams>,
    },
    "/s/:query": {
      id: '/s/:query',
      params: {query: string | number},
      query: ExportedQuery<import('../app/routes/s.$query').SearchParams>,
    },
    "/sign-in/*": {
      id: '/sign-in/*',
      params: {"*": string | number},
      query: ExportedQuery<import('../app/routes/sign-in.$').SearchParams>,
    },
    "/somewhere/cool/*": {
      id: '/somewhere/cool/*',
      params: {"*": string | number},
      query: ExportedQuery<import('../app/catchall').SearchParams>,
    },
    "/static/home": {
      id: '/static/home',
      params: never,
      query: ExportedQuery<import('../app/routes/(static).home').SearchParams>,
    }
  }

  type RoutesWithParams = Pick<
    Routes,
    {
      [K in keyof Routes]: Routes[K]["params"] extends Record<string, never> ? never : K
    }[keyof Routes]
  >;

  export type RouteId = Routes[keyof Routes]['id'];

  export function $path<
    Route extends keyof Routes,
    Rest extends {
      params: Routes[Route]["params"];
      query?: Routes[Route]["query"];
    }
  >(
    ...args: Rest["params"] extends Record<string, never>
      ? [route: Route, query?: Rest["query"]]
      : [route: Route, params: Rest["params"], query?: Rest["query"]]
  ): string;

  export function $params<
    Route extends keyof RoutesWithParams,
    Params extends RoutesWithParams[Route]["params"]
  >(
      route: Route,
      params: { readonly [key: string]: string | undefined }
  ): {[K in keyof Params]: string};

  export function $routeId(routeId: RouteId): RouteId;
}"
`;
